os/20201103-full-disk-raid.md

```diff
@@ -0,0 +1,208 @@
+
+Thus the ESP won't be RAIDed; we'll create multiple independent filesystems with identical directory trees.  Keeping each replica independent avoids RAID desynchronization if the firmware chooses to modify the ESP, and makes sense because CoreOS does not modify the ESP after installation.  Since there will no longer be a single canonical ESP, we'll no longer automount `/boot/efi` in the running system.
+
+When updating the bootloader, bootupd will need to find each partition with an ESP type GUID and update it independently.  This assumes that all ESPs on the system are controlled by CoreOS.
+
+## BIOS Boot partition
+
+The Dracut glue will save an image of the BIOS boot partition whenever the config specifies a partition with the BIOS boot type GUID, and will restore that image to _every_ such partition.  In addition, the same logic will save the boot sector (the first 440 bytes of the disk) and restore it to every disk where we write a BIOS Boot partition.
+
+The BIOS Boot partition is also not updated at runtime, so RAID is not necessary here either.  As a result, when updating the bootloader, bootupd will need to find each BIOS Boot type GUID and update both the partition and the corresponding boot sector.  Again this assumes that the entire system belongs to CoreOS.
+
+For simplicity, we're copying the BIOS Boot partition and boot sector bit-for-bit, rather than rerunning `grub-install` from the initrd or hand-modifying the boot sector.  Since the boot sector hardcodes the offset of the BIOS Boot partition, the latter must be recreated at the same offset as the original partition.  To avoid hardcoding an awkward constant in FCCT, we'll move the BIOS boot partition from its current offset of 512 MiB to the beginning of the disk (offset 1 MiB) in new images.  The Dracut glue will detect relocation of the BIOS boot partition and fail the boot.
+
+There's no BIOS boot partition in 4Kn images.  We don't know in advance which disks will be 4Kn, so the Ignition config will always create a BIOS boot partition on x86_64, but the Dracut glue will automatically skip copying it (and the boot sector) if there's no source partition to copy from.
+
+## PReP partition
+
+For ppc64le, we'll create independent replicas of the PReP partition, similar to the BIOS boot partition.  We'll need to confirm with multi-arch folks that this will work.
+
+## Ignition changes
+
+The Dracut glue can detect the requisite Ignition config stanzas using `jq`, but that's brittle and we're already doing a lot of it in the initrd.  Ideally Ignition will provide command-line arguments for executing relevant (hard-coded) queries against the cached Ignition config.
+
+## Architecture support
+
+We'll need to get feedback from the multi-arch team for aarch64 and ppc64le.  ppc64le will need an alternate `layout` definition that includes the PReP partition.
```

Comment:
i think this would work. The only thing I remember being special about the Prep partition was it needs to be the first partition to work correctly on some systems such as powerVM.